[Product] Update product tag store to migrate Core Data usage and fix a few issues

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #14091
Closes #14513 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR updates the usage of storage in `ProductTagStore`:
- Replace the deprecated usage of `writerDerivedStorage` with the new method `performAndSave` for upserting product tags.
- Handle removing unused tags as part of the upsertion.
- Move `deleteStoredProductTags` to be executed in the background.
- Replace multiple fetch requests with a single one to optimize performance.

While testing the above changes, I found that even though we remove unused tags upon syncing, those tags are still displayed when selecting tags for products. The reason is that we are adding all the tags fetched from the remote to storage again eventually.

I fixed this by parsing the `count` property for product tags and use it to filter only used tags when syncing. This ensures that only used tags are recommended, similar to how the feature works in WooCommerce Core.

Also: addressed #14513 by not triggering the tag creation endpoint when the tag list is empty.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Log in to a store with existing products.
- Create a new product or open an existing one.
- Add a new tag that is not used for any other products and save the changes.
- Open another product and select the tag list. Confirm that the new tag is suggested for the product.
- Open the initial product and select the tag list. Clear all the added tags and save the changes.
- Open another product and select the tag list. Confirm that the new tag is no longer suggested.
- Also: confirm that the tag list is displayed with cached items upon loading.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
Tested and confirmed with simulator iPhone 16 Pro iOS 18.1:
- Product list is loaded correctly.
- Tags are listed correctly for existing products.
- Add/edit products: only used tags are suggested.
- Removing all references to a tag hides it from the suggested list for products.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/ea4c9e99-50d7-4af5-a402-9699523611ec


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Replace deprecated use of writerDerivedStorage

itsmeichigo committed:
Remove unused tags as part of the upsertion

itsmeichigo committed:
Handle deleteStoredProductTags in background

itsmeichigo committed:
Remove unused method deleteUnusedStoredProductTags and move the logic to be as part of upsertion

itsmeichigo committed:
Add new field count for ProductTag

itsmeichigo committed:
Only upsert product tags with associated products

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Fix test build failure

itsmeichigo committed:
Add test for addProductTags to check when tag list is empty

itsmeichigo committed:
Set default value for count to avoid unnecessary changes

itsmeichigo committed:
Fix test failure for ProductTagStore

itsmeichigo committed:
Add comment explaining the `count` property for ProductTag

itsmeichigo committed:
Update unit tests for ProductTagStore

itsmeichigo committed:
Remove count property from tag initializers

itsmeichigo committed:
Merge branch 'trunk' into task/14091-product-tag-store-update

itsmeichigo committed:
Update release notes

itsmeichigo committed:
Fetch all items in one go to avoid multiple fetches

hichamboushaba commented on Yosemite/Yosemite/Stores/ProductTagStore.swift:
```diff
@@ -80,11 +76,12 @@ private extension ProductTagStore {
-
-                self?.upsertStoredProductTagsInBackground(productTags, siteID: siteID) {
+                let shouldRemoveUnusedTags = pageNumber == Default.firstPageNumber
+                /// Only upsert tags that are associated with at least 1 product.
+                let filteredTags = productTags.filter { $0.count > 0 }
```
> I fixed this by parsing the count property for product tags and use it to filter only used tags when syncing. This ensures that only used tags are recommended, similar to how the feature works in WooCommerce Core.

@itsmeichigo I didn't manage to review the rest of the PR yet, but I'm just leaving this remark to get it discussed, in Core, I noticed that tags with `count = 0` are still suggested, as you can see from the following screenshots:

<img width="937" alt="Screenshot 2024-11-27 at 17 08 41" src="https://github.com/user-attachments/assets/29106754-7491-45a5-8aa7-e23e363ba101">
<img width="280" alt="Screenshot 2024-11-27 at 17 08 55" src="https://github.com/user-attachments/assets/f9090293-d344-4258-9a44-66dc88d7aa89">

Can you double check on your side? I'm not sure what's different between our testing environments

itsmeichigo replied:
What I meant by the tags not being suggested is that it's not listed below the text field (similar to when you select Choose from the most used tags). If you think it's not necessary and that we should display them regardless, I can revert the change. The code to remove unused tags in the storage would be redundant and need to be removed as well.

hichamboushaba replied:
> What I meant by the tags not being suggested is that it's not listed below the text field

I'm not sure I understand what you mean here? the tags listed under the text field in Core are the ones that are already applied to this product, or do you mean something else.

> similar to when you select Choose from the most used tags

Personally I think the app's UX is more aligned with the text field of core than this button, we don't offer a similar thing.

> If you think it's not necessary and that we should display them regardless, I can revert the change.

I would vote to do this yes, but it's not a blocker for the PR, so I'll leave the final decision to you. And just for information, the Android app suggests those tags too, so if we revert this change, we'll keep all platforms aligned.

> The code to remove unused tags in the storage would be redundant and need to be removed as well.

True.

------ 
One related point, if you end up reverting the code, maybe we can sort the tags by their count in the app, this way I think we'll have a UX that offers an experience similar to what Core offers with its widget (I mean the fact that it has an autocomplete text field in addition to the "Choose from the most used tags" button), WDYT? This doesn't need to be done on this PR though.

itsmeichigo replied:
Sorting tags by count sound like a good idea - we'd need to persist the count of products for each tag in the storage to do that. For now, I've reverted the irrelevant changes to make it easier to review this PR - please take another look and let me know what you think - thanks!

itsmeichigo committed:
Merge branch 'trunk' into task/14091-product-tag-store-update

itsmeichigo committed:
Revert change to synchronizeProductTags method

itsmeichigo committed:
Revert addition of count property

itsmeichigo committed:
Revert changes to unit tests

hichamboushaba commented on Yosemite/Yosemite/Stores/ProductTagStore.swift:
```diff
@@ -130,32 +126,19 @@ private extension ProductTagStore {
+        }, completion: onCompletion, on: .main)
     }
 
     /// Deletes any Storage.ProductTag with the specified `siteID` and `productID`
```
np, the function takes the `ids` of the product tags, and not `productID`, we should update the comment.

itsmeichigo replied:
Good catch! I updated the comment in 31008d341b40ecd3dafd29b89584edc328c035f9.

hichamboushaba commented on Yosemite/Yosemite/Stores/ProductTagStore.swift:
```diff
@@ -96,6 +88,9 @@ private extension ProductTagStore {
     /// Create new product tags associated with a given Site ID.
     ///
     func addProductTags(siteID: Int64, tags: [String], onCompletion: @escaping (Result<[ProductTag], Error>) -> Void) {
+        guard tags.isEmpty == false else {
+            return onCompletion(.success([]))
+        }
```
I found it interesting that we need this check here, after checking the code, the issue is in the handling of tags in iOS, the logic in `ProductTagsViewController` could use some optimization, as currently it always tries to create tags even the already added ones, and by extension causes the issue that you noticed.
Normally, the logic should be create just the new tags, and would speed up the process also (when the user picks just existing ones).

I'll create an issue to keep track of this improvement; for now, this fix here should be enough.

itsmeichigo replied:
Thank you!

itsmeichigo replied:
Since this is a simple fix, I sneaked it in c77d166c09054466c27e2937c0b708024be077d2.

hichamboushaba replied:
Thanks @itsmeichigo for the additional improvement, this looks good, but my remark touches another point, currently the logic of iOS always triggers an API request to create the tags, but this should be needed only when the user adds a new one that doesn't already exist, when the user picks tags from the suggestions, we don't need to re-create them. The following screen recording shows what I mean (I'm selecting an existing tag, and yet we show a loading indicator and we send an API request):

https://github.com/user-attachments/assets/4e57e9f3-c68f-4d25-8688-9796eff4b9e7

I created this ticket https://github.com/woocommerce/woocommerce-ios/issues/14555 to keep track of this.

hichamboushaba reviewed (approved):
Nice work @itsmeichigo, all works as expected (tested using iPhone 16 simulator, iOS 18.1)

One point though, it seems the cached data is not used anywhere, we always fetch tags from remote before showing them, I think a good improvement to do to this screen is to update it to show the cached data first, WDYT? I can create a ticket for this improvement if you agree.

itsmeichigo committed:
Update comment on deleteStoredProductTags

itsmeichigo committed:
Display cached tags immediately upon loading tags

itsmeichigo committed:
Only trigger tag creation if there are any tags confirmed

itsmeichigo committed:
Update release notes

itsmeichigo commented:
Thanks @hichamboushaba for the helpful feedback! 

> One point though, it seems the cached data is not used anywhere, we always fetch tags from remote before showing them, I think a good improvement to do to this screen is to update it to show the cached data first, WDYT?

Nice observation! This is a simple fix also so I added an update in cd4946a3efcf76f846f6c7cb782120a1179d4dbe.

I also updated the release notes and PR description to take into account the new update. Please let me know if there is anything else needs updating.
